### ISSUE TITLE
`coqfind` design.

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#7312]
Current organization of Coq libraries a bit adhoc, with them being managed on a load-path basis and making it quite hard for projects to manage any kind of dependencies.

In order to improve this situation, we should improve `coqdune` to have the notion of an installed Coq library in the fashion `ocamlfind` works. Thus, people can declare `((libraries coquelicot))` and have things work automatically.

jsCoq already implements the beginning of a Coq package format, but we won't evolve a lot on this issue until at least #6857 is merged.

**Comment by Zimmi48** [comment]
Could you summarize a bit more concretely what kind of problems are encountered with the load path based management?

Additionally can you tell more about the preliminary Coq package format you have? It seems that this question could possibly interact with the redesign of the Coq package index that @maximedenes has in mind.

**Comment by ejgallego** [comment]
The idea is that a `coq` package is a mapping from a name to a set of options, usually load paths in the case of Coq. There is nothing like this now, so it cannot be compared.

Our current "package" management system is based on dropping everything into a single directory, usually `user-contribs`, and the recursively scanning such directories and adding all the contents blindly to the loadpath.

It doesn't take a lot of effort to see all the problems this scheme has when your number of packages is large, for example, a user cannot easily tell Coq _not_ to use a package.

The preliminary package format I implemented for jscoq is very easy, a JSON file which for every package lists the packages it depends and lists the necessary loadpaths to be added to Coq when the package is needed.

So in my view, a Coq development should add a `coqbuild` file (to replace `_CoqProject`) and that file would include stuff such as `((libraries mathcomp.algebra stdpp))`, etc... then, all the files in the dir get to use that _exact_ set of packages.

**Comment by maximedenes** [comment]
> The preliminary package format I implemented for jscoq is very easy, a JSON file which for every package lists the packages it depends and lists the necessary loadpaths to be added to Coq when the package is needed.

That sounds very good. It would be natural to try to use this format as the primary notion of a Coq package descriptor, and generate OPAM files / Nix derivations from it for the Coq Package Index.

**Comment by ejgallego** [comment]
> That sounds very good. It would be natural to try to use this format as the primary notion of a Coq package descriptor, and generate OPAM files / Nix derivations from it for the Coq Package Index.

Way more work is needed but indeed the tool that generates the json files already has the knowledge to generate the `.install` files [which is the tricky part].

A nice side-effect of doing things this way is that you don't care whether the library you are using is installed system-wide or locally, as the loadpath for Coq is properly setup reading the `coqbuild` file and understanding what's available under the current working copy.

**Comment by maximedenes** [comment]
@ejgallego It would be nice to write a short document explaining the format we want for Coq packages. Is that on your todo list?

**Comment by ejgallego** [comment]
I am a bit short on time now to comment more, but I think we want to merge Dune first, switch OPAM and most CI to Dune, then we should try to put some package using `coqfind`.

IMHO the design is yet too open to know, also Dune is not close to 1.0 so we need to wait to see what the final set of features will be as to know how to best integrate.

Also, we are waiting for feedback from @RalfJung as to know his feature wanted list.

**Comment by RalfJung** [comment]
I should preface that I am not familiar with dune at all, and hardly familiar with OCaml. I am looking at this from the perspective of someone who only uses Coq's own "surface language" (i.e., I do Ltac and the like, but no OCaml plugins). Also I am heavily influenced by Rust and it's build system Cargo because I think they're doing a great job :D

From what I gather, there are multiple goals mixed in this discussion. One is to replace the manual loadpath handling we have currently. Here the assumption is that something else installed everything and now we just need to find out where it was installed. Such a tool is not even necessary currently as everything is installed in a well-known location, so is the goal to be more flexible here? Doing this doesn't need much information, so I guess just having the names of the dependencies suffices -- though one may want to get nice error messages if the dependency is tool old, which obviously would require a version number.

But then you also mention *generating* the opam file. So it seems there are some plans to replace the entire dependency declaration with `coqfind`, both for installing and for linking against dependencies. That's actually pretty similar to what the `Cargo.toml` file does in Rust land. There, the common case is to depend on a package that was uploaded at https://crates.io/. So for example to use the [regex](https://crates.io/crates/regex) crate I would add a line in the `[dependencies]` section saying
```
regex = "1.0.0"
```
The version number here is a semver constraint, so this will match any version `1.x.y`. I imagine in Coq land this could refer to packages uploaded into the [opam repo](https://coq.inria.fr/opam/www/) (except that AFAIK opam does not understand semver so expressing these constraints would be harder -- but then actually satisfying semver in Coq is pretty hard, so that could be for the better).

However, for the in-development version of our libraries, we don't want to depend just on released versions (that would mean releasing every other day). But we also don't want to just say `iris = "dev"` as that makes it impossible to figure out which version of Iris to use. It is pretty much impossible to avoid breaking e.g. LambdaRust when updating Iris, so always using the latest will not work reliably. Moreover, if we go back to a LambdaRust from 2 months ago, how should we figure out which Iris version to use? So, instead we'd like to depend on a development version of Iris.

Cargo actually supports this: Instead of giving a version number and letting cargo figure out from https://crates.io where to get the package, you can just tell it directly where to get the package:
```
# depend on another package in the same repo
hello_utils = { path = "hello_utils" }
# depend on a package in another repo (using the latest master branch)
my-library = { git = 'https://example.com/git/my-library' }
# depend on a particular rev, tag or branch
my-library = { git = 'https://example.com/git/my-library', rev = '460c58b9978988117268aa69c3a4192d4da4018c' }
my-library = { git = 'https://example.com/git/my-library', tag = 'v0.8~beta' }
my-library = { git = 'https://example.com/git/my-library', branch = 'some-feature' }
```
Those git repositories will contain Rust packages, so in particular they will contain a `Cargo.toml` defining *their* dependencies. Thus Cargo can make sure all packages have their dependencies installed in the exact version they need. You can find more details in the [cargo documentation](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html).

So, that's essentially the feature I wish I had for Iris development: Being able to point the package manager directly at a git repository, avoiding the round-trip through a version released in some package manager.

Now, building Rust programs works pretty different from building Coq programs: Cargo actually compiles these dependencies independently for every project. So if I have some Rust application in `~/src/my-app`, then all its dependencies will be built in `~/src/my-app/target` and not interfere with the dependencies installed for other applications elsewhere. But OTOH, opam now supports "local switches" which can give you a pretty similar experience, so maybe doing something similar for Ciq is not entirely unrealistic.

**Comment by vbgl** [comment]
@ejgallego wrote:

> Our current "package" management system is based on dropping everything into a single directory, usually user-contribs, and the recursively scanning such directories and adding all the contents blindly to the loadpath.

This refers to your own experience. Here is what I do: every library is dropped in a distinct `user-contrib` directory (corollary: no two libraries end up in the same `user-contrib` directory at install-time). Then I use one of these two different tricks: a) either gather all libraries I need under a single `user-contrib` directory by means of symlinks; or b) set the `COQPATH` environment variable to the exact list of directories I need. I personally prefer option (b) but that’s mostly a matter of taste.

Of course, I have to be careful about 1) library dependencies and 2) with which version of Coq which library is compiled with.

Doing this kind of management by hand might be tedious: this is why we have package managers. I believe most package managers should be able to deal with dependencies and symlinks/env. variables. I would be surprised if new formats/tools/… were actually needed.

**Comment by ejgallego** [comment]
> This refers to your own experience.

This refers to the experience of everybody not overriding the default `coq_makefile` behavior; which I do believe is the large majority of users.

`COQPATH` or similar is fine however it is not correct to say that this is exclusively the job of the package manager; it is the build system who should bring into scope the correct dependencies, and only those for the build. In your scheme [I guess Nix-based] if I happen to have in my environment 10 Coq libraries all of them will be visible by `coqtop`.

That's not OK IMO, and note that in many cases you have parts of the same package depending on different libs so the mapping cannot be made 1:1.

**Comment by RalfJung** [comment]
Another good reason for using some kind of `coqfind` instead of `COQPATH` just occurred to me: Right now, I have 3 opam switches, each with their own copy of Coq, because I want to have different versions of Iris available for different projects. It would be nice if, instead of having completely separate switches, I could install multiple versions of Iris inside the same switch and have different projects use the same Coq, but different iris.

**Comment by ejgallego** [comment]
> Another good reason for using some kind of coqfind instead of COQPATH just occurred to me: Right now, I have 3 opam switches, each with their own copy of Coq, because I want to have different versions of Iris available for different projects. It would be nice if, instead of having completely separate switches, I could install multiple versions of Iris inside the same switch and have different projects use the same Coq, but different iris.

Umm, `COQPATH` + installing each version of Iris on their own dir seems like way to go here, doesn't it?

**Comment by ejgallego** [comment]
> installing each version of Iris on their own dir

I am not sure if Coq makefile supports this but it should I think.

**Comment by RalfJung** [comment]
I'd have to do all that completely manually though. Right now I just do `make build-dep` and that invokes opam to make my current switch have the version of Iris that I need.

**Comment by ejgallego** [comment]
You make a good point on having versioning; that's not something Dune / Ocamlfind handle, and it is for a reason, but maybe we should.

**Comment by ejgallego** [comment]
Ok, so I am going to implement a prototype PR as follows:
- I will add a new directory, `coqfind` to the sources. This directory will provide the `coqfind` library and will be able to bootstrap separately from Coq.
- `coqfind` will export an API similar to http://projects.camlcity.org/projects/dl/findlib-1.8.0/doc/ref-html/lib/Findlib.html , namely an `init`, `list_packages`, and `package_property` methods. However, most things will be statically typed.
- `coqfind` may support several methods to locate packages, the idea is to use path variables and maybe a dune-inspired automatic setup.
- a package will be a directory specifiying a `coqpackage` file. This file will use sexp syntax and will declare the name of the package, dependencies on other packages, and filesystem structure. This file could indeed contain the build instructions too, so in fact, we could just directly write a `dune` file and make it act as a package descriptor. This could be the long-term plan.
- whether coq should depend on `coqfind` or `coqfind` on Coq is an interesting question. I lean towards the first option, so Coq can understand package natively. This makes it harder to use coqfind to locate Coq tho, as it is done in `ocamlfind`.

All the code in Coq doing traversal of the file system would be deprecated. Coq will only know how to add things to the load path as instructed by `coqfind`.

**Comment by Zimmi48** [comment]
> This file will use sexp syntax and will declare the name of the package, dependencies on other packages, and filesystem structure.

I won't find over this but I'd personally prefer if it used JSON syntax instead of sexp syntax. It would be more standard and more readable.

> This file could indeed contain the build instructions too, so in fact, we could just directly write a dune file and make it act as a package descriptor. This could be the long-term plan.

Obviously, if you want this file to be a correct Dune file, then you have to follow the sexp syntax they have chosen.

**Comment by SkySkimmer** [comment]
sexp is the one that's more readable though.
And considering dune uses sexps it's also locally more standard.

**Comment by Zimmi48** [comment]
> sexp is the one that's more readable though

First time, I hear that LISP-like syntax is readable. But anyways, as I said, I won't fight for it.

**Comment by ejgallego** [comment]
> I won't find over this but I'd personally prefer if it used JSON syntax instead of sexp syntax. It would be more standard and more readable.

Hi @Zimmi48 , I recommend reading https://github.com/ocaml/dune/issues/801 and https://github.com/ocaml/dune/issues/461#issuecomment-391741822

Note that I first chose JSON for the package format, however the main advantage that IMHO sexp have is that Dune does provide us a lexer/parser so using it is one dependency less for `coqfind`.

**Comment by ejgallego** [comment]
Finally the plan here has changed, and we will use native Dune support as suggested by Dune devs.

A first prototype of the library format is available at ocaml/dune#1555.

The discussion will continue on #8052 and ocaml/dune#1555, thus closing.

**Comment by ejgallego** [comment]
@RalfJung note that Dune plans to support similar features to what you have requested, once the Coq Dune PR is merged we should move the discussion there.
